notebooks/network-analysis.ipynb

```diff
@@ -0,0 +1,434 @@
+    "- `reverse_network` flips the orientation of the network such the tailwaters are at the head of the network, and connections propagate into the headwaters. In doing so, each segments upstream connections are identified. The function accepts the `connections` dictionary, and returns a dictionary of segments and their respective upstream connections. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Given the network DataFrame, `extract_network` returns a dictionary segments and downstream connections.\n",
+    "# segment-id: [downstream-connected-segment-id]\n",
+    "connections = nhd_network.extract_network(data, network_data[\"downstream_col\"])\n",
+    "\n",
+    "# flip the orientation of the network from headwaters -> tailwaters, to tailwaters -> headwaters\n",
+    "# this allows us to enumerate each segment's upstream neighbor\n",
+    "# segment-id: [upstream-connected-segment-id]\n",
+    "rconn = nhd_network.reverse_network(connections)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Identify subnetworks\n",
+    "- The `reachable` function breaks the supernetwork into constituent subnetworks by extracting all segments connected to each tailwater segment. \n",
+    "- The function accepts the reverse network dictionary, `rconn` (created in the cell above), and returns a dictionary, `subreachable`, containing a list of segments with connections to each tailwater segment in the supernetwork.\n",
+    "- Example output is shown for a single tailwater segment"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# determine all of the stream segments with upstream connections to each tail water (terminal node) in the supernetwork\n",
+    "# note we hand the `reachable` function the reversed network dictionary `rconn`\n",
+    "# tailwater-segment-id: [list of ALL upstream connected segment ids]\n",
+    "subreachable = nhd_network.reachable(rconn)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Visualize subnetworks\n",
+    "- Make a map showing each of the subnetworks in the supernetwork with different colors. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Get the tailwater keys for each subnetwork \n",
+    "n_subnets = subreachable.keys()\n",
+    "\n",
+    "# initialize a new column in the GeoDataFrame that will containa subnetwork code\n",
+    "gdf[\"subnetwork\"] = 0\n",
+    "\n",
+    "# loop through subnetworks, populate the newly created \"subnetwork\" columm with an integer subnetwork identifier code\n",
+    "i = 1\n",
+    "for sub in n_subnets:\n",
```

Comment:
This loop can be structured better (and probably faster too):
```python
for i, reached in enumerate(subreachable.values(), 1):
    gdf.loc[gdf.index.isin(reached), "subnetwork"] = i
```
The negative nodes shouldn't be appearing in the subreachable lists. They get filtered out by nhd_network.extract_network.